---
title: "Alignment-free expression estimation using Sailfish"
author: "Radhika Khetani"
date: ""
---

Contributors: Radhika Khetani, Meeta Mistry

Approximate time: 3 hours

## Learning Objectives

* Understand the concept of "Pseudocounts"
* Understand how to use Sailfish to generate Pseudocounts
* Learn how to perform differential gene expression on Pseudocounts

## What is Sailfish

[Sailfish](http://www.cs.cmu.edu/~ckingsf/software/sailfish/index.html) and it's more recent "upgrade" [Salmon](https://combine-lab.github.io/salmon/), are based on the philosophy of lightweight algorithms. They use the sequence of genes or transcripts as input, and do not align the whole read. Instead it's a 2-step process based on counting "kmers":

**a.** they first evaluate the sequences for all possible unique sequences of length k (kmer) in the transcriptome (genes/transcripts).

**b.** then they count the number of times those kmers appear in the sequenced data, i.e. the fastq. This count information is then used to estimate the abundance of each gene or transcript. 

<img src="../img/nbt.2862-F1.jpg" width="400">

## Running Sailfish

    mkdir ~/ngs_course/rnaseq/sailfish
    cd ~/ngs_course/rnaseq/sailfish
    export PATH=/groups/bcbio/bcbio/anaconda/bin:/opt/bcbio/local/bin:$PATH
    
As you can imagine from the above schematic, there are 2 steps in the analysis too:

a. "Index" the transcriptome (transcripts or genes) as follows:
    
    #sailfish index sailfish index -p <num of cores> -k <kmer size> -t <fasta of gene sequences> -o <folder name>

**We are not going to run this in class, but it only takes a few minutes.** We will be using an index we have generated from transcript sequences (all known transcripts with multiples for some genes), but this can be generated from genic sequences too. 

b. Get the abundance using the quantification step as follows:

    sailfish quant -i /groups/hbctraining/sailfish-run/sailfish.ensembl2.idx/ \
    -l SR \
    -r ngs_course/rnaseq/data/untrimmed_fastq/Mov10_oe_1.subset.fq \
    --useVBOpt \
    -o Mov10_oe_1.subset.sailfish
    
    less Mov10_oe_1.subset.sailfish/quant.sf
  
You can run this command on all the datasets and then use the quant tables in R for the next few steps. We have already performed these runs for you and you can download the directory with all the necessary files [from here](). Once you download it, please do the following:
* Decompress the tar archive
* Open RStudio
* Go to the files tab and find your way to the new folder you downloaded
* Set that folder your working directory

## Converting Sailfish output to pseudo-counts and performing DE analysis

The pseudocounts generated by sailfish are represented as normalized TPM (transcripts per million) counts and can represent transcripts or genes, depending on the input in the index step. These need to be converted into non-normalized count estimates for performing DESeq2 analysis. In addition, we need to be able to input these values with decimal places to create the DESeq object. 

### Setting up to run DESeq2 on pseudocount data:

The developers of DESeq2 are aware of the abovem-entioned issues, so they:
* have developed a package that can make this conversion ([tximport](https://github.com/mikelove/tximport#update-now-hosted-on-bioconductor))
* have also created a new custom function that enables DESeq2 to accept decimal places in count values to generate a DESeq object

The `tximport` package available on bioconductor is currently too new and will not work with our regular R installation. This only temporary and will change after April 2016. In the meantime, we have to grab an older version from github, and to download packages from github, we need a new package called `devtools`.

**Step 1.** Install all the new packages required (you'll only need to do this once)
    
    install.packages("devtools")

    library(devtools)   # required for installing any packages from github
    
    install_github("mikelove/tximport")         # installing from github

    source("DESeqDataFromTx.R") # required for using tximport output as input for DESeq2

    install.packages("readr")       # required to generate the output from tximport

    source("http://bioconductor.org/biocLite.R")
    biocLite("biomaRt")             # tximport requires gene symbols as row names
    
**Step 2.** Load the libraries required for setting up

    library(tximport)
    library(readr)
    library(biomaRt)

**Step 3.** Load the data from the working directory (which is the folder we downloaded earlier)
    
    dir <- getwd()
    
    samples <- scan(file.path(dir,"samples.txt"), what="character")
    
    files <- file.path(dir, samples, "quant.sf")
    
    names(files) <-  sapply(files, function(x){strsplit(x, "/")[[1]][5]}, USE.NAMES=F)

**Step 4.** Create a data frame containing Ensembl Transcript IDs and Gene symbols

Our Sailfish index was generated with transcript sequences, but tximport needs to know which Genes these transcripts came from, so we need to use the `biomaRt` package to extract this information.

    head(read.delim(files[1], sep="\t", header=T))
    
    ids <- read.delim(files[1], sep="\t", header=T)
    ids <- as.character(ids[,1])
   

**NOTE a: Do not run the next section of code**, this code is for future reference only. The `getBM()` function takes a long time to run, so we have already generated the information the `tx2gene` data frame needs to have. *Mary will be going over the `getBM` function in more detail in the Ensembl/Biomart lesson.*
    
    # mart<- useDataset("hsapiens_gene_ensembl", useMart("ENSEMBL_MART_ENSEMBL", host="www.ensembl.org"))
    # gene.names <- getBM(
    #    filters= "ensembl_transcript_id", 
    #     attributes= c("ensembl_gene_id", "external_gene_name", "ensembl_transcript_id"),
    #    values= ids,
    #    mart= mart)
    # tx2gene <- gene.names[,3:2]
    
**NOTE b: Run the following code in class**, but remember that if you obtain the values from biomaRt directly, you won't have to run it.

    tx2gene <- read.table("tx2gene.txt", header=T)
    
**Step 5:** Run tximport to summarize gene-level information    
  
    txi <- tximport(files, type="sailfish", txIn = TRUE, txOut = FALSE, tx2gene=tx2gene, reader=read_tsv)
    colnames(txi$counts) <- samples
  
### Using DESeq2 for DE analysis with pseudocounts
    
    library(DESeq2)
    source('DESeqDataFromTx.R')
    
    sampleTable <- data.frame(condition=factor(c(rep("Ctl",3), rep("KD", 2), rep("OE", 3))))
    
    rownames(sampleTable) <- colnames(txi$counts)
    
    dds <- DESeqDataSetFromTximport(txi, sampleTable, ~ condition)
    
Now you have created a DESeq object to proceed with DE analysis as we discussed in the last session!

***
*This lesson has been developed by members of the teaching team at the [Harvard Chan Bioinformatics Core (HBC)](http://bioinformatics.sph.harvard.edu/). These are open access materials distributed under the terms of the [Creative Commons Attribution license](https://creativecommons.org/licenses/by/4.0/) (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited.*
